### ISSUE TITLE
Disallow passing non integer indices to jax.nn.one_hot

### ISSUE THREAD

**Issue by bunelr** [jax-ml/jax#15882]
When evaluating the expression `jax.nn.one_hot(jnp.array([0.2, 1.2, 2.]), 3)`,

the result is 

```
Array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 1.]], dtype=float32)
```

because the first two indices are not integer, so will not match any coordinates.

Would it be worthwhile  to raise an error when the indices argument is not an integer type, to prevent silent errors in case someone forgot to cast their indices?

**Comment by jakevdp** [comment]
Thanks for pointing this out – I think you're right that it would make sense to require input indices to be integers, particularly given the confusing behavior when non-integers are supplied.

**Comment by jakevdp** [comment]
So I've dug-in a bit and this is not going to be a straightforward change: there are many downstream users of this utility who pass in floating-point values, some very deliberately (e.g. using `NaN` as a marker of rows for which there is no "hot" index).

Even if we were to fix these explicit cases, there is a chance that changing this behavior would change production models that implicitly rely on the current behavior.

Given that, I'm thinking that the more reasonable way forward would be to document and test the behavior of this function for non-integer inputs, without changing the function's behavior. What do you think?

**Comment by bunelr** [comment]
That's true that this would definitely be a breaking change (and one that I imagine is quite hard to fix automatically, given that whether or not casting will change anything depends on the value contained in the tensor). I guess the `NaN` usage could be replaced by using negative values for the same behaviour, but other use-case may not be as straightforward.

Flagging the problem in the documentation sounds like a good compromise :)

**Comment by bunelr** [comment]
It seems that since #25590, the problematic behaviour is now deprecated (and warnings get raised).
Closing the issue!
